pkg/core/marshal/marshal.go

```diff
@@ -0,0 +1,187 @@
+package marshal
+
+import (
+	"reflect"
+
+	coreReflect "github.com/skygeario/skygear-server/pkg/core/reflect"
+)
+
+// UpdateNilFieldsWithZeroValue checks the fields with tag
+// `default_zero_value:"true"` and updates the fields with zero value if they are nil.
+// This function will walk through the struct recursively,
+// if the tagged fields of struct have duplicated type in the same path,
+// the function may cause infinite loop.
+// Before calling this function, please make sure the struct get pass with
+// function `shouldNotHaveDuplicatedTypeInSamePath` in the test case.
+func UpdateNilFieldsWithZeroValue(i interface{}) {
+	t := reflect.TypeOf(i).Elem()
+	v := reflect.ValueOf(i).Elem()
+
+	if t.Kind() != reflect.Struct {
+		return
+	}
+	numField := t.NumField()
+	for i := 0; i < numField; i++ {
+		zerovalueTag := t.Field(i).Tag.Get("default_zero_value")
+		if zerovalueTag != "true" {
+			continue
+		}
+
+		field := v.Field(i)
+		ft := t.Field(i)
+		if field.Kind() == reflect.Ptr {
+			ele := field.Elem()
+			if !ele.IsValid() {
+				ele = reflect.New(ft.Type.Elem())
+				field.Set(ele)
+			}
+			UpdateNilFieldsWithZeroValue(field.Interface())
+		}
+	}
+}
+
+// OmitEmpty traverses the given struct and set pointer to struct to nil if
+// the value IsRecursivelyZero.
+func OmitEmpty(i interface{}) {
```

Comment:
Where will we use the `OmitEmpty` function?

Reply:
In controller